File: apps/demo/modules/expo-module-showcase/ios/ExpoModuleShowcaseModule.swift
import ExpoModulesCore
import UIKit

public final class ExpoModuleShowcaseModule: Module {
  private var counts = ["created": 0, "foregrounded": 0, "backgrounded": 0, "startObserving": 0, "stopObserving": 0]
  private let lock = NSLock()
  private var sequence = 0
  private var page: ShowcasePage?

  public func definition() -> ModuleDefinition {
    Name("ExpoModuleShowcase")
    Constants(["platform": "ios", "nativeLanguage": "Swift"])
    Events("onShowcaseEvent")
    Function("echo") { (value: String) in value }
    AsyncFunction("echoAsync") { (value: String) in value }
    AsyncFunction("failAsync") { () throws -> Void in
      throw Exception(name: "ShowcaseException", description: "这是用于测试的原生错误。", code: "ERR_SHOWCASE")
    }
    Function("emitEvent") { (value: String) in
      self.sequence += 1
      self.sendEvent("onShowcaseEvent", ["value": value, "sequence": self.sequence, "platform": "ios"])
    }
    Class(ShowcaseSharedCounter.self) {
      Constructor { (value: Double) in ShowcaseSharedCounter(value) }
      Property("value") { (counter: ShowcaseSharedCounter) in counter.value }
        .set { (counter: ShowcaseSharedCounter, value: Double) in counter.value = value }
      Function("increment") { (counter: ShowcaseSharedCounter, delta: Double) in counter.increment(delta) }
      AsyncFunction("incrementAsync") { (counter: ShowcaseSharedCounter, delta: Double) in counter.increment(delta) }
      Function("emitValueChanged") { (counter: ShowcaseSharedCounter) in
        counter.emit(event: "onValueChanged", arguments: ["value": counter.value])
      }
    }
    Class(ShowcaseTextRef.self) {
      Property("value") { (ref: ShowcaseTextRef) in ref.ref }
    }
    Function("returnSameSharedCounter") { (counter: ShowcaseSharedCounter) in counter }
    Function("createSharedTextRef") { (value: String) in ShowcaseTextRef(value) }
    View(ShowcaseView.self) {
      Events("onValueChanged")
      Prop("label") { (view: ShowcaseView, label: String) in view.label = label }
      Prop("value") { (view: ShowcaseView, value: Double) in view.value = value }
      AsyncFunction("increment") { (view: ShowcaseView, delta: Double) in view.increment(delta, source: "command") }
    }
    AsyncFunction("openNativePage") { (initialValue: Double, promise: Promise) in
      guard self.page == nil, let presenter = self.appContext?.utilities?.currentViewController(),
        presenter.viewIfLoaded?.window != nil, !presenter.isBeingDismissed else {
        throw Exception(name: "ShowcasePageException", description: "没有可用的页面容器，或原生页面已经打开。", code: "ERR_SHOWCASE_PAGE")
      }
      let page = ShowcasePage(initialValue: initialValue) { [weak self] action, value in
        self?.page = nil
        promise.resolve(["action": action, "value": value])
      }
      self.page = page
      let navigation = UINavigationController(rootViewController: page)
      navigation.modalPresentationStyle = .fullScreen
      navigation.overrideUserInterfaceStyle = .light
      navigation.navigationBar.tintColor = ShowcaseStyles.blue
      let appearance = UINavigationBarAppearance()
      appearance.configureWithOpaqueBackground()
      appearance.backgroundColor = ShowcaseStyles.canvas
      navigation.navigationBar.standardAppearance = appearance
      navigation.navigationBar.scrollEdgeAppearance = appearance
      presenter.present(navigation, animated: true)
    }.runOnQueue(.main)
    Function("getLifecycleSnapshot") { () -> [String: Int] in
      self.lock.lock()
      defer { self.lock.unlock() }
      return self.counts
    }
    OnCreate { self.record("created") }
    OnAppEntersForeground { self.record("foregrounded") }
    OnAppEntersBackground { self.record("backgrounded") }
    OnStartObserving { self.record("startObserving") }
    OnStopObserving { self.record("stopObserving") }
    OnDestroy {
      DispatchQueue.main.async { self.page?.finish(action: "cancel"); self.page = nil }
    }
  }

  private func record(_ name: String) {
    lock.lock()
    defer { lock.unlock() }
    counts[name, default: 0] += 1
  }
}

File: apps/demo/modules/expo-module-showcase/android/src/main/java/expo/modules/showcase/ExpoModuleShowcaseModule.kt
package expo.modules.showcase

import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.atomic.AtomicInteger

class ExpoModuleShowcaseModule : Module() {
  private val created = AtomicInteger()
  private val foregrounded = AtomicInteger()
  private val backgrounded = AtomicInteger()
  private val startObserving = AtomicInteger()
  private val stopObserving = AtomicInteger()
  private val sequence = AtomicInteger()
  private var page: ShowcasePage? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoModuleShowcase")
    Constants("platform" to "android", "nativeLanguage" to "Kotlin")
    Events("onShowcaseEvent")
    Function("echo") { value: String -> value }
    AsyncFunction("echoAsync") { value: String -> value }
    AsyncFunction("failAsync") { promise: Promise ->
      promise.reject("ERR_SHOWCASE", "这是用于测试的原生错误。", null)
    }
    Function("emitEvent") { value: String ->
      sendEvent("onShowcaseEvent", mapOf("value" to value, "sequence" to sequence.incrementAndGet(), "platform" to "android"))
    }
    Class(ShowcaseSharedCounter::class) {
      Constructor { value: Double -> ShowcaseSharedCounter(appContext, value) }
      Property("value") { counter: ShowcaseSharedCounter -> counter.value }
        .set { counter: ShowcaseSharedCounter, value: Double -> counter.value = value }
      Function("increment") { counter: ShowcaseSharedCounter, delta: Double -> counter.increment(delta) }
      AsyncFunction("incrementAsync") { counter: ShowcaseSharedCounter, delta: Double -> counter.increment(delta) }
      Events("onValueChanged")
      Function("emitValueChanged") { counter: ShowcaseSharedCounter ->
        counter.emit("onValueChanged", mapOf("value" to counter.value))
      }
    }
    Class(ShowcaseTextRef::class) {
      Property("value") { ref: ShowcaseTextRef -> ref.ref }
    }
    Function("returnSameSharedCounter") { counter: ShowcaseSharedCounter -> counter }
    Function("createSharedTextRef") { value: String -> ShowcaseTextRef(appContext, value) }
    View(ShowcaseView::class) {
      Events("onValueChanged")
      Prop("label") { view: ShowcaseView, label: String -> view.label = label }
      Prop("value") { view: ShowcaseView, value: Double -> view.value = value }
      AsyncFunction("increment") { view: ShowcaseView, delta: Double -> view.increment(delta, "command") }
    }
    AsyncFunction("openNativePage") { initialValue: Double, promise: Promise ->
      if (page != null) throw CodedException("ERR_SHOWCASE_PAGE", "原生页面已经打开。", null)
      val activity = appContext.currentActivity
        ?: throw CodedException("ERR_SHOWCASE_PAGE", "没有可显示页面的 Activity。", null)
      if (activity.isFinishing || activity.isDestroyed) {
        throw CodedException("ERR_SHOWCASE_PAGE", "Activity 已销毁。", null)
      }
      val nativePage = ShowcasePage(activity, initialValue)
      page = nativePage
      nativePage.setOnDismissListener {
        if (page === nativePage) page = null
        promise.resolve(mapOf("action" to nativePage.action, "value" to nativePage.value))
      }
      try {
        nativePage.show()
      } catch (error: Exception) {
        page = null
        throw error
      }
    }.runOnQueue(Queues.MAIN)
    Function("getLifecycleSnapshot") {
      mapOf("created" to created.get(), "foregrounded" to foregrounded.get(), "backgrounded" to backgrounded.get(),
        "startObserving" to startObserving.get(), "stopObserving" to stopObserving.get())
    }
    OnCreate { created.incrementAndGet() }
    OnActivityEntersForeground { foregrounded.incrementAndGet() }
    OnActivityEntersBackground { backgrounded.incrementAndGet() }
    OnStartObserving { startObserving.incrementAndGet() }
    OnStopObserving { stopObserving.incrementAndGet() }
    OnActivityDestroys { closePage() }
    OnDestroy { closePage() }
  }

  private fun closePage() {
    Handler(Looper.getMainLooper()).post { page?.dismiss(); page = null }
  }
}
